cmd/vicadmin/server.go

```diff
@@ -112,39 +127,208 @@ func (s *server) listenPort() int {
 	return s.l.Addr().(*net.TCPAddr).Port
 }
 
+// Enforces authentication on route `link` and runs `handler` on successful auth
+func (s *server) AuthenticatedHandle(link string, h http.Handler) {
+	s.Authenticated(link, h.ServeHTTP)
+}
+
+func (s *server) Handle(link string, h http.Handler) {
+	s.mux.Handle(link, gorillacontext.ClearHandler(h))
+}
+
+// Enforces authentication on route `link` and runs `handler` on successful auth
+func (s *server) Authenticated(link string, handler func(http.ResponseWriter, *http.Request)) {
+	defer trace.End(trace.Begin(""))
+
+	authHandler := func(w http.ResponseWriter, r *http.Request) {
+		websession, _ := s.uss.cookies.Get(r, sessionCookieKey) // ignore error because it is okay if it doesn't exist
+
+		if len(r.TLS.PeerCertificates) > 0 { // the user is authenticated by certificate at connection time
+			usersess := s.uss.Add("root", &rootConfig.Config)
+
+			timeNow, err := usersess.created.MarshalText()
+			if err != nil {
+				// it's probably safe to ignore this error since we just created usersess.created when we called Add() above
+				// but just in case..
+				log.Errorf("Failed to unmarshal time object %+v into text due to error: %s", usersess.created, err)
+				http.Error(w, genericErrorMessage, http.StatusInternalServerError)
+				return
+			}
+
+			websession.Values[sessionCreationTimeKey] = string(timeNow)
+			websession.Values[sessionUsernameKey] = "root"
+			err = websession.Save(r, w)
+			if err != nil {
+				log.Errorf("Could not create session for user authenticated via client certificate due to error \"%s\"", err.Error())
+				http.Error(w, genericErrorMessage, http.StatusInternalServerError)
+				return
+			}
+
+			// user was authenticated via cert
+			handler(w, r)
+			return
+		}
+
+		c := websession.Values[sessionCreationTimeKey]
+		if c == nil { // no cookie, so redirect to login
+			log.Errorf("No authentication token: %+v", websession.Values)
+			http.Redirect(w, r, loginPagePath, http.StatusTemporaryRedirect)
+			return
+		}
+
+		// here we have a cookie, but we need to make sure it's not expired:
+		// parse the cookie creation time
+		created, err := time.Parse(time.RFC3339, c.(string))
+		if err != nil {
```

Comment:
If that's the solution, could we have a "Click here to log in again" or similar?